File: app/src/main/java/dev/ragnarok/fenrir/activity/SelectionUtils.java
package dev.ragnarok.fenrir.activity;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.model.FavePage;
import dev.ragnarok.fenrir.model.Owner;
import dev.ragnarok.fenrir.model.SelectProfileCriteria;
import dev.ragnarok.fenrir.model.User;
import dev.ragnarok.fenrir.util.Logger;
import dev.ragnarok.fenrir.util.Utils;

public class SelectionUtils {

    private static final String TAG = SelectionUtils.class.getSimpleName();
    private static final String VIEW_TAG = "SelectionUtils.SelectionView";

    public static void addSelectionProfileSupport(Context context, ViewGroup root, Object mayBeUser) {
        if (!(context instanceof ProfileSelectable) || root == null) return;

        SelectProfileCriteria criteria = ((ProfileSelectable) context).getAcceptableCriteria();

        boolean canSelect = false;
        if (criteria.getIsPeopleOnly() ? mayBeUser instanceof User : mayBeUser instanceof Owner || mayBeUser instanceof FavePage) {
            canSelect = true;
        }

        if (canSelect && criteria.getOwnerType() == SelectProfileCriteria.OwnerType.ONLY_FRIENDS) {
            assert mayBeUser instanceof User;
            canSelect = ((User) mayBeUser).isFriend();
        }

        ProfileSelectable callack = (ProfileSelectable) context;
        ImageView selectionView = root.findViewWithTag(VIEW_TAG);

        if (!canSelect && selectionView == null) return;

        if (canSelect && selectionView == null) {
            selectionView = new ImageView(context);
            selectionView.setImageResource(R.drawable.plus);
            selectionView.setTag(VIEW_TAG);
            selectionView.setBackgroundResource(R.drawable.circle_back);
            selectionView.getBackground().setAlpha(150);
            selectionView.setLayoutParams(createLayoutParams(root));

            int dp4px = (int) Utils.dpToPx(4, context);
            selectionView.setPadding(dp4px, dp4px, dp4px, dp4px);

            Logger.d(TAG, "Added new selectionView");
            root.addView(selectionView);
        } else {
            Logger.d(TAG, "Re-use selectionView");
        }

        selectionView.setVisibility(canSelect ? View.VISIBLE : View.GONE);

        if (!canSelect) {
            selectionView.setOnClickListener(null);
        } else {
            if (mayBeUser instanceof FavePage && ((FavePage) mayBeUser).getOwner() != null) {
                selectionView.setOnClickListener(v -> callack.select(((FavePage) mayBeUser).getOwner()));
            } else if (mayBeUser instanceof Owner) {
                selectionView.setOnClickListener(v -> callack.select((Owner) mayBeUser));
            }
        }
    }

    private static ViewGroup.LayoutParams createLayoutParams(ViewGroup parent) {
        if (parent instanceof FrameLayout) {
            int margin = (int) Utils.dpToPx(6, parent.getContext());

            FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

            params.bottomMargin = margin;
            params.leftMargin = margin;
            params.rightMargin = margin;
            params.topMargin = margin;
            return params;
        } else {
            throw new IllegalArgumentException("Not yet impl for parent: " + parent.getClass().getSimpleName());
        }
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/activity/SelectionUtils.kt
package dev.ragnarok.fenrir.activity

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.model.FavePage
import dev.ragnarok.fenrir.model.Owner
import dev.ragnarok.fenrir.model.SelectProfileCriteria
import dev.ragnarok.fenrir.model.User
import dev.ragnarok.fenrir.util.Logger
import dev.ragnarok.fenrir.util.Utils

object SelectionUtils {
    private val TAG = SelectionUtils::class.java.simpleName
    private const val VIEW_TAG = "SelectionUtils.SelectionView"


    fun addSelectionProfileSupport(context: Context?, root: ViewGroup?, mayBeUser: Any) {
        if (context !is ProfileSelectable || root == null) return
        val criteria = (context as ProfileSelectable).acceptableCriteria
        var canSelect =
            if (criteria?.getIsPeopleOnly() == true) mayBeUser is User else mayBeUser is Owner || mayBeUser is FavePage
        if (canSelect && criteria?.getOwnerType() == SelectProfileCriteria.OwnerType.ONLY_FRIENDS) {
            assert(mayBeUser is User)
            canSelect = (mayBeUser as User).isFriend
        }
        val callack = context as ProfileSelectable
        var selectionView =
            root.findViewWithTag<ImageView>(VIEW_TAG)
        if (!canSelect && selectionView == null) return
        if (canSelect && selectionView == null) {
            selectionView = ImageView(context)
            selectionView.setImageResource(R.drawable.plus)
            selectionView.tag = VIEW_TAG
            selectionView.setBackgroundResource(R.drawable.circle_back)
            selectionView.background.alpha = 150
            selectionView.layoutParams = createLayoutParams(root)
            val dp4px = Utils.dpToPx(4f, context).toInt()
            selectionView.setPadding(dp4px, dp4px, dp4px, dp4px)
            Logger.d(TAG, "Added new selectionView")
            root.addView(selectionView)
        } else {
            Logger.d(TAG, "Re-use selectionView")
        }
        selectionView?.visibility = if (canSelect) View.VISIBLE else View.GONE
        if (!canSelect) {
            selectionView.setOnClickListener(null)
        } else {
            if (mayBeUser is FavePage && mayBeUser.owner != null) {
                mayBeUser.owner?.let { vv ->
                    selectionView.setOnClickListener { callack.select(vv) }
                }
            } else if (mayBeUser is Owner) {
                selectionView.setOnClickListener {
                    callack.select(
                        mayBeUser
                    )
                }
            }
        }
    }

    private fun createLayoutParams(parent: ViewGroup): ViewGroup.LayoutParams {
        return if (parent is FrameLayout) {
            val margin = Utils.dpToPx(6f, parent.getContext()).toInt()
            val params = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
            )
            params.bottomMargin = margin
            params.leftMargin = margin
            params.rightMargin = margin
            params.topMargin = margin
            params
        } else {
            throw IllegalArgumentException("Not yet impl for parent: " + parent.javaClass.simpleName)
        }
    }
}